### ISSUE TITLE
SPIKE: Custom implementation vs using Sidekiq for PushmiPullyu App

### ISSUE THREAD

**Issue by murny** [ualbertalib/pushmi_pullyu#17]
If you look at the workflow documented here https://github.com/ualbertalib/pushmi_pullyu/issues/16, this is pretty much exactly how [Sidekiq](https://github.com/mperham/sidekiq) works. PushmiPullyu is essentially a background job processor.

- Jobs are enqueued in a redis queue (after save callback that pushes GF into redis queue)
- Sidekiq montiors and retrieves jobs from redis queue (what PushmiPullyu will do)
- Performs job (retrieve data from fedora, builds AIP, bags AIP, pushes AIP to Swift, and logs events)

Do we need to custom implement everything for PushmiPullyu? At very least we should be using Sidekiq and other job queues as ideas.

Pros for using Sidekiq: 

- Work for running PushmiPullyu on a server as a CLI and dameonizing it will not be required
- Proven and stable track record for running in production
- Fantastic Logs and built in Dashboard
- Easier to maintain and for new people to get trained on (most rails devs know how to use job queues)
- Sidekiq is thread based and concurrent (could be a concern too?)

Things to be concerned about regarding Sidekiq:

- Sidekiq does not support unique jobs out of the box, requires addons, etc. (pro version also has unique jobs)
  - https://github.com/mhenrixon/sidekiq-unique-jobs
- Time window period logic could be done when enqueuing the job from ERA (`PushmiPullyuJob.set().perform later()`)

Henry asks about replay support for failed jobs (sidekiq has this built in, how would we tackle this using a custom approach)?

**Comment by mbarnett** [comment]
Here's what the raw redis version looks like

Populating a Redis priority set (the Hydranorth `after_save` side of things, is just a call to `zadd`:
```ruby
 :001 > r = Redis.new
 => #<Redis client v3.3.3 for redis://127.0.0.1:6379/0> 
 :002 > 10.times do |i|
 :003 >     r.zadd 'pmpy_queue', Time.now.to_f, "noid#{i}"
 :004?>   end
 => 10 
 :005 > r.zrange('pmpy_queue', 0, -1)
 => ["noid0", "noid1", "noid2", "noid3", "noid4", "noid5", "noid6", "noid7", "noid8", "noid9"] 
```

The Pushmi-Pullyu side can be seen in this PR. A sample interaction with it looks like:

```ruby
 ./bin/console
 :001 > q = PushmiPullyu::PreservationQueue.new
 => #<PushmiPullyu::PreservationQueue:0x007fd2561c1a30 @redis=#<Redis client v3.3.3 for redis://127.0.0.1:6379/0>> 
 :002 > q.wait_next_item
 => "noid0" 
 :003 > q.wait_next_item
 => "noid1" 
# <etc>
# elements return immediately until the queue is empty, at which point
# wait_next_item blocks until there is something in the queue
```

**Comment by mbarnett** [comment]
* I first investigated ways of achieving unique, priority ordered queue using raw Redis infrastructure (using sorted sets in Redis), as seen in this experimental PR: https://github.com/ualbertalib/pushmi_pullyu/pull/41

* After that I looked at the feasability of using Sidekiq. sidekiq-unique-job doesn't have easy support for evicting an earlier job if a later one comes along with the same unique parameters. It wants the first job in to "own" the lock on that parameter and only unlock the uniqueness constraint once the job has executed. We could modify this, but I think it's more work than its worth.

* Upshot: we can do this pretty easily using just 3 Redis commands, and we can isolate this in a way that should make it trivial to replace with some other queueing mechanism (an RDBMS, rabbit, w/e) if that proves unsatisfactorily simple. The Sidekiq approach requires writing more code *and* tying our implementation to Sidekiq's job architecture.

* Currently thinking about thinking about ways of ensuring replayability if a noid fails

**Comment by murny** [comment]
We chose to go with a custom implementation approach as Sidekiq didn't quite exactly fit our needs. Would have been alot of customization and forcing sidekiq to do what we wanted. Easier to just build what we want.

With regards to "Currently thinking about thinking about ways of ensuring replayability if a noid fails", matt says: 

`I think it’s two fold. On the HN side, something like skylight or honeybadger to report that it couldn’t get the noid into Redis, and on the PMPY side, logging it in pairs of starting on noid 1/finished archiving noid1, so that we can detect noids that came out of the queue but didn’t finish. We could also store the noid in a separate queue during processing, but that’s still vulnerable to any kind of connection issue with Redis. Any kind of simple replayability of noids like that is probably out of scope for the initial crack at it, though.` 

So make sure we capture this stuff with logs and alerts/monitoring ( #32 ) and any type of automated replayability of noids will be out of scope for MVP.

Closing out issue.
